packages/vite/src/node/server/transformRequest.ts

```diff
@@ -179,17 +198,20 @@ async function doTransform(
     }
   }
 
-  if (ssr) {
-    return (mod.ssrTransformResult = await ssrTransform(
-      code,
-      map as SourceMap,
-      url
-    ))
-  } else {
-    return (mod.transformResult = {
-      code,
-      map,
-      etag: getEtag(code, { weak: true })
-    } as TransformResult)
+  const result = ssr
+    ? await ssrTransform(code, map as SourceMap, url)
+    : ({
+        code,
+        map,
+        etag: getEtag(code, { weak: true })
+      } as TransformResult)
+
+  // Only cache the result if the module wasn't invalidated while it was
+  // being processed, so it is re-processed next time if it is stale
+  if (timestamp > mod.lastInvalidationTimestamp) {
+    if (ssr) mod.ssrTransformResult = result
+    else mod.transformResult = result
   }
```

Comment:
For the invalidation step, I wonder if we should move this up to

https://github.com/vitejs/vite/blob/db17172fbf50dce2cdd026e410d46f9430f1424a/packages/vite/src/node/server/transformRequest.ts#L71-L82

So that we compare the timestamp when there's a cache.

I'm not familiar with module invalidation and such, but I saw `doTransform` [strips the timestamp query](https://github.com/vitejs/vite/blob/db17172fbf50dce2cdd026e410d46f9430f1424a/packages/vite/src/node/server/transformRequest.ts#L62), and I'm not sure if it's something we can re-use or if it's an entirely different thing.

Reply:
Talked with patak earlier, and the suggestion I proposed could be risky since the cache could be overwritten in certain race conditions. So not saving the cache after it's been identified stale is the safest path forward.

Reply:
We talked about the PRs in a call, resolving this conversation as we think the current fix is ok.